rllib/models/torch/fcnet.py

```diff
@@ -142,11 +142,16 @@ def forward(
         seq_lens: TensorType,
     ) -> (TensorType, List[TensorType]):
         obs = input_dict["obs_flat"].float()
-        self._last_flat_in = obs.reshape(obs.shape[0], -1)
-        self._features = self._hidden_layers(self._last_flat_in)
-        logits = self._logits(self._features) if self._logits else self._features
-        if self.free_log_std:
-            logits = self._append_free_log_std(logits)
+        # TODO: use the correct type for CUDA
+        with torch.autocast(
```

Comment:
the torch version we use in CI doesn't have autocast, it seems.

Reply:
Oops! It should. I think in `pytorch-1.10` they alias `torch.cuda.amp.autocast` to `torch.autocast`. I will use the longer version instead. I don't think in the older versions in works on CPU, but it's kinda pointless to use it on CPU anyways.

Reply:
For `torch-1.9` as listed in https://github.com/ray-project/ray/blob/53c4c7b1be54fe67a54b34f3f0c076ad502e35cc/python/requirements_ml_docker.txt#L14-L15: https://pytorch.org/docs/1.9.0/amp.html#torch.cuda.amp.autocast

Reply:
cool. let's see if the tests pass after it.

also a really clueless question, why is this controlled by obs.is_cuda bit, and not say, the actual nn model?

Reply:
So devices in torch are only defined for `Tensors`. `torch.nn.Modules` do not have a device. This is because maybe you want some of your layers on the CPU and some on the GPU. It can be annoying, because often your model is only on one device, but that's the price of flexibility I suppose.

Instead, we can just check what device the tensors are on. If we try and multiply `obs.cuda() * torch.tensor(1.).cuda()` we'll get an error that `obs` and `1` are on different devices. So we can be pretty sure that the `obs` device is the same as all the parameters in the model.

Reply:
interesting. thanks for the explanation.